datafusion/src/optimizer/remove_duplicate_filters.rs

```diff
@@ -0,0 +1,611 @@
+                _ => expr.clone(),
+            })
+            .unwrap_or(expr.clone()),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::Or,
+            right,
+        } if expr_contains(right, left) => as_binary_expr(right)
+            .map(|x| match x {
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::Or,
+                    right: _,
+                } => *right.clone(),
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::And,
+                    right: _,
+                } => *left.clone(),
+                _ => expr.clone(),
+            })
+            .unwrap_or(expr.clone()),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if expr_contains(left, right) => as_binary_expr(left)
+            .map(|x| match x {
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::Or,
+                    right: _,
+                } => *right.clone(),
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::And,
+                    right: _,
+                } => x.clone(),
+                _ => expr.clone(),
+            })
+            .unwrap_or(expr.clone()),
+        Expr::BinaryExpr {
+            left,
+            op: Operator::And,
+            right,
+        } if expr_contains(right, left) => as_binary_expr(right)
+            .map(|x| match x {
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::Or,
+                    right: _,
+                } => *left.clone(),
+                Expr::BinaryExpr {
+                    left: _,
+                    op: Operator::And,
+                    right: _,
+                } => x.clone(),
+                _ => expr.clone(),
+            })
+            .unwrap_or(expr.clone()),
+        Expr::BinaryExpr { left, op, right } => Expr::BinaryExpr {
+            left: Box::new(simplify(&left)),
+            op: *op,
+            right: Box::new(simplify(right)),
+        },
+        _ => expr.clone(),
+    }
+}
+
+fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
+    match plan {
+        LogicalPlan::Filter { input, predicate } => Ok(LogicalPlan::Filter {
+            input: input.clone(),
+            predicate: simplify(predicate),
+        }),
+        LogicalPlan::Projection { expr, input, schema } => Ok(LogicalPlan::Projection {
+            expr: expr.into_iter().map(|x| simplify(x)).collect::<Vec<_>>(),
+            input: input.clone(),
+            schema: schema.clone(),
+        }),
+        _ => {
+            let new_inputs = plan
+                .inputs()
+                .iter()
+                .map(|input| optimize(input))
+                .collect::<Result<Vec<_>>>()?;
+
+            let expr = plan.expressions();
```

Comment:
This could use something like `plan.expressions().into_iter().map(|x| simplify(x)).collect::<Vec<_>>()` ? and the other patterns could be removed here? (So it also applies for any node containing expressions.

Reply:
Should this be valid ?

```
fn optimize(plan: &LogicalPlan) -> Result<LogicalPlan> {
    let new_inputs = plan
        .inputs()
        .iter()
        .map(|input| optimize(input))
        .collect::<Result<Vec<_>>>()?;
    let expr = plan
        .expressions()
        .into_iter()
        .map(|x| simplify(&x))
        .collect::<Vec<_>>();
    utils::from_plan(&plan, &expr, &new_inputs)
}
```

Reply:
That looks right to me!